File: pkg/service/auth_service.go
package service

import (
	"errors"
	"time"

	"teaching-management/pkg/config"
	"teaching-management/pkg/model"
	"teaching-management/pkg/repository"

	"github.com/golang-jwt/jwt/v5"
	"github.com/google/uuid"
	"golang.org/x/crypto/bcrypt"
)

type AuthService struct {
	userRepo *repository.UserRepository
}

func NewAuthService(userRepo *repository.UserRepository) *AuthService {
	return &AuthService{userRepo: userRepo}
}

func (s *AuthService) Register(req *model.RegisterRequest) (*model.AuthResponse, error) {
	// Check if user already exists
	existingUser, err := s.userRepo.GetByEmail(req.Email)
	if err != nil {
		return nil, err
	}
	if existingUser != nil {
		return nil, errors.New("email already registered")
	}

	// Hash password
	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(req.Password), bcrypt.DefaultCost)
	if err != nil {
		return nil, err
	}

	user := &model.User{
		Name:         req.Name,
		Email:        req.Email,
		PasswordHash: string(hashedPassword),
	}

	if err := s.userRepo.Create(user); err != nil {
		return nil, err
	}

	// Generate tokens
	token, refreshToken, err := s.GenerateTokens(user.ID)
	if err != nil {
		return nil, err
	}

	return &model.AuthResponse{
		Token:        token,
		RefreshToken: refreshToken,
		User:         *user,
	}, nil
}

func (s *AuthService) Login(req *model.LoginRequest) (*model.AuthResponse, error) {
	user, err := s.userRepo.GetByEmail(req.Email)
	if err != nil {
		return nil, err
	}
	if user == nil {
		return nil, errors.New("invalid email or password")
	}

	// Compare passwords
	err = bcrypt.CompareHashAndPassword([]byte(user.PasswordHash), []byte(req.Password))
	if err != nil {
		return nil, errors.New("invalid email or password")
	}

	// Generate tokens
	token, refreshToken, err := s.GenerateTokens(user.ID)
	if err != nil {
		return nil, err
	}

	return &model.AuthResponse{
		Token:        token,
		RefreshToken: refreshToken,
		User:         *user,
	}, nil
}

func (s *AuthService) RefreshToken(tokenStr string) (string, string, error) {
	// Parse refresh token
	token, err := jwt.Parse(tokenStr, func(t *jwt.Token) (interface{}, error) {
		if _, ok := t.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, errors.New("unexpected signing method")
		}
		return []byte(config.AppConfig.JWTRefreshSecret), nil
	})

	if err != nil || !token.Valid {
		return "", "", errors.New("invalid refresh token")
	}

	claims, ok := token.Claims.(jwt.MapClaims)
	if !ok {
		return "", "", errors.New("invalid token claims")
	}

	sub, ok := claims["sub"].(string)
	if !ok {
		return "", "", errors.New("invalid token subject")
	}

	userID, err := uuid.Parse(sub)
	if err != nil {
		return "", "", errors.New("invalid user ID in token")
	}

	// Re-generate both tokens
	return s.GenerateTokens(userID)
}

func (s *AuthService) GenerateTokens(userID uuid.UUID) (string, string, error) {
	cfg := config.AppConfig

	// Access Token
	accessClaims := jwt.MapClaims{
		"sub": userID.String(),
		"exp": time.Now().Add(time.Duration(cfg.JWTAccessExpiryMin) * time.Minute).Unix(),
		"iat": time.Now().Unix(),
	}
	accessToken := jwt.NewWithClaims(jwt.SigningMethodHS256, accessClaims)
	accessTokenStr, err := accessToken.SignedString([]byte(cfg.JWTSecret))
	if err != nil {
		return "", "", err
	}

	// Refresh Token
	refreshClaims := jwt.MapClaims{
		"sub": userID.String(),
		"exp": time.Now().Add(time.Duration(cfg.JWTRefreshExpiryDays) * 24 * time.Hour).Unix(),
		"iat": time.Now().Unix(),
	}
	refreshToken := jwt.NewWithClaims(jwt.SigningMethodHS256, refreshClaims)
	refreshTokenStr, err := refreshToken.SignedString([]byte(cfg.JWTRefreshSecret))
	if err != nil {
		return "", "", err
	}

	return accessTokenStr, refreshTokenStr, nil
}

File: src/services/auth.service.ts
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import { config } from '../config/index.js';
import { UserRepository } from '../repositories/interfaces.js';
import { RegisterRequest, LoginRequest, AuthResponse } from '../types/index.js';

export class AuthService {
  constructor(private userRepo: UserRepository) {}

  async register(req: RegisterRequest): Promise<AuthResponse> {
    const existing = await this.userRepo.getByEmail(req.email);
    if (existing) {
      throw new Error('email already registered');
    }

    const salt = await bcrypt.genSalt(10);
    const passwordHash = await bcrypt.hash(req.password, salt);

    const user = await this.userRepo.create({
      name: req.name,
      email: req.email,
      password_hash: passwordHash
    });

    const tokens = this.generateTokens(user.id);

    return {
      token: tokens.token,
      refresh_token: tokens.refreshToken,
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        created_at: user.created_at,
        updated_at: user.updated_at
      }
    };
  }

  async login(req: LoginRequest): Promise<AuthResponse> {
    const user = await this.userRepo.getByEmail(req.email);
    if (!user || !user.password_hash) {
      throw new Error('invalid email or password');
    }

    const isMatch = await bcrypt.compare(req.password, user.password_hash);
    if (!isMatch) {
      throw new Error('invalid email or password');
    }

    const tokens = this.generateTokens(user.id);

    return {
      token: tokens.token,
      refresh_token: tokens.refreshToken,
      user: {
        id: user.id,
        name: user.name,
        email: user.email,
        created_at: user.created_at,
        updated_at: user.updated_at
      }
    };
  }

  async refreshToken(tokenStr: string): Promise<{ token: string; refresh_token: string }> {
    try {
      const decoded = jwt.verify(tokenStr, config.jwtRefreshSecret) as { sub: string };
      if (!decoded || !decoded.sub) {
        throw new Error('invalid token subject');
      }

      const user = await this.userRepo.getById(decoded.sub);
      if (!user) {
        throw new Error('user not found');
      }

      const tokens = this.generateTokens(user.id);
      return {
        token: tokens.token,
        refresh_token: tokens.refreshToken
      };
    } catch (err: any) {
      throw new Error('invalid refresh token');
    }
  }

  generateTokens(userId: string): { token: string; refreshToken: string } {
    const accessToken = jwt.sign(
      { sub: userId },
      config.jwtSecret,
      { expiresIn: `${config.jwtAccessExpiryMin}m` }
    );

    const refreshToken = jwt.sign(
      { sub: userId },
      config.jwtRefreshSecret,
      { expiresIn: `${config.jwtRefreshExpiryDays}d` }
    );

    return { token: accessToken, refreshToken };
  }
}
